src/org/rascalmpl/repl/TerminalProgressBarMonitor.java

```diff
@@ -0,0 +1,459 @@
+     */
+    private final String encoding;
+
+    /**
+     * Will make everything slow, but easier to spot mistakes
+     */
+    private final boolean debug = false;
+
+    @SuppressWarnings("resource")
+    public TerminalProgressBarMonitor(OutputStream out, Terminal tm) {
+        super(out);
+        this.encoding = tm.getOutputEncoding() != null ? tm.getOutputEncoding() : "UTF8";
+        PrintWriter theWriter = new PrintWriter(out, true, Charset.forName(encoding));
+        this.writer = debug ? new PrintWriter(new AlwaysFlushAlwaysShowCursor(theWriter)) : theWriter;
+        this.lineWidth = tm.getWidth();
+
+        assert System.console() != null: "interactive progress bar needs a terminal, and we should not print this into a file anyway.";
+    }
+
+    /**
+     * Use this for debugging terminal cursor movements, step by step.
+     */
+    private static class AlwaysFlushAlwaysShowCursor extends FilterWriter {
+
+        public AlwaysFlushAlwaysShowCursor(PrintWriter out) {
+            super(out);
+        }
+
+        @Override
+        public void write(int c) throws IOException {
+            out.write(c);
+            out.write(ANSI.showCursor());
+            out.flush();
+        }
+
+        @Override
+        public void write(char[] cbuf, int off, int len) throws IOException {
+            out.write(cbuf, off, len);
+            out.write(ANSI.showCursor());
+            out.flush();
+        }
+
+        @Override
+        public void write(String str, int off, int len) throws IOException {
+            out.write(str, off, len);
+            out.write(ANSI.showCursor());
+            out.flush();
+        }
+
+       
+    }
+
+    /**
+     * Represents one currently running progress bar
+     */
+    private class ProgressBar {
+        private final long threadId;
+        private final String threadName;
+        private final String name;
+        private int max;
+        private int current = 0;
+        private int previousWidth = 0;
+        private int doneWidth = 0;
+        private final int barWidth = lineWidth - "☐ ".length() - " 🕐 00:00:00.000 ".length();
+        private final Instant startTime;
+        private Duration duration;
+        private String message = "";
+        private int stepper = 1;
+        private final String[] clocks = new String[] {"🕐" , "🕑", "🕒", "🕓", "🕔", "🕕", "🕖", "🕗", "🕘", "🕙", "🕛"};
+        public int nesting = 0;
+
+        ProgressBar(String name, int max) {
+            this.threadId = Thread.currentThread().getId();
+            this.threadName = Thread.currentThread().getName();
+            this.name = name;
+            this.max = Math.max(1, max);
+            this.startTime = Instant.now();
+            this.duration = Duration.ZERO;
+            this.message = name;
+        }
+
+        void worked(int amount, String message) {
+            this.current += Math.min(amount, max);
```

Comment:
can't this still run past `max`? shouldn't it be `this.current = Math.min(current + amount, max)`?